File: app/src/main/java/com/sunflower/pantaucovid19/ui/fragment/BeritaFragment.java
package com.sunflower.pantaucovid19.ui.fragment;

import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import com.bumptech.glide.Glide;
import com.frogobox.frogonewsapi.data.model.Article;
import com.frogobox.frogonewsapi.data.response.ArticleResponse;
import com.frogobox.frogonewsapi.util.NewsConstant;
import com.frogobox.frogonewsapi.util.NewsUrl;
import com.frogobox.recycler.FrogoRecyclerView;
import com.frogobox.recycler.callback.FrogoAdapterCallback;
import com.sunflower.pantaucovid19.R;
import com.sunflower.pantaucovid19.base.BaseFragment;
import com.sunflower.pantaucovid19.source.DataRepository;
import com.sunflower.pantaucovid19.source.remote.GetRemoteCallback;
import com.sunflower.pantaucovid19.ui.activity.WebViewActivity;

import org.jetbrains.annotations.NotNull;

import java.util.List;
import java.util.Objects;

/**
 * A simple {@link Fragment} subclass.
 */
public class BeritaFragment extends BaseFragment {
    private SwipeRefreshLayout swipe2refresh;
    private ProgressBar progressBar;
    private FrogoRecyclerView frogoRecyclerView;

    public BeritaFragment() {
        // Required empty public constructor
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        return inflater.inflate(R.layout.fragment_berita, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        frogoRecyclerView = view.findViewById(R.id.rv_news_category);
        progressBar = view.findViewById(R.id.progress_view);
        swipe2refresh = view.findViewById(R.id.refreshNews);
        getTopHeadLine(frogoRecyclerView, progressBar);
        swipeAction();
    }

    private void swipeAction() {
        swipe2refresh.setOnRefreshListener(() -> {
            getTopHeadLine(frogoRecyclerView, progressBar);
            swipe2refresh.setRefreshing(false);
        });
        swipe2refresh.setColorSchemeResources(R.color.colorPrimary);
    }

    private void getTopHeadLine(FrogoRecyclerView frogoRecyclerView, ProgressBar progressBar) {
        DataRepository dataRepository = new DataRepository(getContext());
        dataRepository.getTopHeadline(
                NewsUrl.NEWS_API_KEY,
                "Virus Corona",
                null,
                null,
                NewsConstant.COUNTRY_ID,
                new GetRemoteCallback<ArticleResponse>() {
                    @Override
                    public void onSuccess(ArticleResponse data) {
                        setupRecyclerView(frogoRecyclerView, data.getArticles());
                    }

                    @Override
                    public void onFailed(String errorMessage) {
                        showToastLong(errorMessage);
                    }

                    @Override
                    public void onShowProgress() {
                        progressBar.setVisibility(View.VISIBLE);
                    }

                    @Override
                    public void onHideProgress() {
                        progressBar.setVisibility(View.GONE);
                    }
                }
        );
    }

    private void setupRecyclerView(FrogoRecyclerView frogoRecyclerView, List<Article> articles) {

        frogoRecyclerView.injectAdapter(
                R.layout.list_article_vertical,
                articles,
                null,
                new FrogoAdapterCallback<Article>() {
                    @Override
                    public void setupInitComponent(@NotNull View view, Article article) {

                        TextView tvTitle = view.findViewById(R.id.tv_title);
                        TextView tvPublishdDate = view.findViewById(R.id.tv_published);
                        TextView tvDescription = view.findViewById(R.id.tv_description);
                        ImageView ivUrl = view.findViewById(R.id.iv_url);

                        tvTitle.setText(article.getTitle());
                        tvPublishdDate.setText(article.getPublishedAt());
                        tvDescription.setText(article.getDescription());
                        Glide.with(view.getContext()).load(article.getUrlToImage()).into(ivUrl);

                    }

                    @Override
                    public void onItemClicked(Article article) {
                        showToastShort(article.getTitle());
                        Intent intent = new Intent(getActivity(), WebViewActivity.class);
                        intent.putExtra("url", article.getUrl());
                        Objects.requireNonNull(getActivity()).startActivity(intent);
                    }

                    @Override
                    public void onItemLongClicked(Article article) {
                        showToastShort(article.getTitle());
                    }
                });
        frogoRecyclerView.isViewLinearVertical(false);
    }


}

File: app/src/main/java/com/sunflower/pantaucovid19/ui/fragment/BeritaFragment.kt
package com.sunflower.pantaucovid19.ui.fragment

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.ProgressBar
import android.widget.TextView
import com.bumptech.glide.Glide
import com.frogobox.coreapi.news.NewsConstant.COUNTRY_ID
import com.frogobox.coreapi.news.NewsUrl
import com.frogobox.coreapi.news.model.Article
import com.frogobox.coreapi.news.response.ArticleResponse
import com.frogobox.recycler.core.FrogoRecyclerNotifyListener
import com.frogobox.recycler.core.IFrogoViewAdapter
import com.frogobox.recycler.widget.FrogoRecyclerView
import com.sunflower.pantaucovid19.R
import com.sunflower.pantaucovid19.base.BaseFragment
import com.sunflower.pantaucovid19.databinding.FragmentBeritaBinding
import com.sunflower.pantaucovid19.source.DataRepository
import com.sunflower.pantaucovid19.source.remote.GetRemoteCallback
import com.sunflower.pantaucovid19.ui.activity.WebViewActivity

/**
 * A simple [Fragment] subclass.
 */
class BeritaFragment : BaseFragment<FragmentBeritaBinding>() {

    override fun setupViewBinding(
        inflater: LayoutInflater,
        container: ViewGroup?
    ): FragmentBeritaBinding {
        return FragmentBeritaBinding.inflate(inflater, container, false)
    }

    override fun setupOnViewCreated(view: View, savedInstanceState: Bundle?) {

        getEverythings(binding.rvNewsCategory, binding.progressView)
        swipeAction()
    }

    private fun swipeAction() {
        binding.apply {
            refreshNews.setOnRefreshListener {
                getEverythings(binding.rvNewsCategory, binding.progressView)
                refreshNews.isRefreshing = false
            }
            refreshNews.setColorSchemeResources(R.color.colorPrimary)
        }

    }

    private fun getEverythings(frogoRecyclerView: FrogoRecyclerView, progressBar: ProgressBar) {
        val dataRepository = DataRepository(context)
        dataRepository.getEverythings(
            NewsUrl.API_KEY,
            "Covid 19",
            null,
            null,
            null,
            null,
            null,
            null,
            COUNTRY_ID,
            null,
            object : GetRemoteCallback<ArticleResponse> {
                override fun onSuccess(data: ArticleResponse) {
                    data.articles?.let { setupRecyclerView(frogoRecyclerView, it) }
                }

                override fun onFailed(errorMessage: String) {
                    showToastLong(errorMessage)
                }

                override fun onShowProgress() {
                    mActivity?.runOnUiThread {
                        progressBar.visibility = View.VISIBLE
                    }
                }

                override fun onHideProgress() {
                    mActivity?.runOnUiThread {
                        progressBar.visibility = View.GONE
                    }
                }
            }
        )
    }

    private fun setupRecyclerView(frogoRecyclerView: FrogoRecyclerView, articles: List<Article>) {
        val frogoViewAdapterCallback = object : IFrogoViewAdapter<Article> {
            override fun setupInitComponent(
                view: View,
                data: Article,
                position: Int,
                notifyListener: FrogoRecyclerNotifyListener<Article>
            ) {
                val tvTitle = view.findViewById<TextView>(R.id.tv_title)
                val tvPublishdDate = view.findViewById<TextView>(R.id.tv_published)
                val tvDescription = view.findViewById<TextView>(R.id.tv_description)
                val ivUrl = view.findViewById<ImageView>(R.id.iv_url)
                tvTitle.text = data.title
                tvPublishdDate.text = data.publishedAt
                tvDescription.text = data.description
                Glide.with(view.context).load(data.urlToImage).into(ivUrl)
            }

            override fun onItemLongClicked(
                view: View,
                data: Article,
                position: Int,
                notifyListener: FrogoRecyclerNotifyListener<Article>
            ) {
                showToastShort(data.title)
            }

            override fun onItemClicked(
                view: View,
                data: Article,
                position: Int,
                notifyListener: FrogoRecyclerNotifyListener<Article>
            ) {
                showToastShort(data.title)
                val intent = Intent(activity, WebViewActivity::class.java)
                intent.putExtra("url", data.url)
                requireActivity().startActivity(intent)
            }
        }
        frogoRecyclerView.injector<Article>()
            .addData(articles)
            .addCustomView(R.layout.list_article_vertical)
            .addEmptyView(null)
            .addCallback(frogoViewAdapterCallback)
            .createLayoutLinearVertical(false)
            .build()
    }

}